### PR TITLE
Add support to set global options

### PR DESCRIPTION
**Pull request by samikshan**
Global options can be set using volume name as 'all'.

No support is there for validation of global option values so far.

Signed-off-by: Samikshan Bairagya <samikshan@gmail.com>

### CODE REVIEW

**Comment by atinmu** [comment]
22:24:18 ./glusterd2/xlator/options/options.go:87:6: exported type GlobalOption should have comment or be unexported
22:24:18 Found 1 lint suggestions; failing.
22:24:18 ./glusterd2/xlator/global.go:20:2: exported var GlobalOptMap should have comment or be unexported
22:24:18 ./glusterd2/xlator/global.go:64:1: exported function LoadGlobalOptions should have comment or be unexported
22:24:18 ./glusterd2/xlator/global.go:65:9: should omit 2nd value from range; this loop is equivalent to `for k := range ...`
22:24:18 Found 3 lint suggestions; failing.
22:24:18 ./glusterd2/xlator/global.go does not conform to gofmt rules. Run: gofmt -s -w ./glusterd2/xlator/global.go
22:24:18 ./glusterd2/xlator/utils.go:10:2: exported const GlobalOptsPrefix should have comment (or a comment on this block) or be unexported
22:24:18 Found 1 lint suggestions; failing.
22:24:19 ./pkg/api/volume_resp.go:74:1: comment on exported type VolumeOptionsGetResp should be of the form "VolumeOptionsGetResp ..." (with optional leading article)
22:24:19 Found 1 lint suggestions; failing.

^^ lint failures.  @samikshan please fix this.

**Comment by kshlm** [comment]
Let's try not to shoehorn global options (the ones that work on volume "all") into the normal volume set API/command. The options being set here don't just operate on volumes, but instead act on the full cluster (eg. cluster.op-version).

I'd really like this to be a seperate API and a seperate command. Ideally, all these would be set on a cluster-object that is persisted in the store.

**Comment by amarts** [comment]
the 'global' option possible with https://review.gluster.org/#/c/18315/ is meant to be more of per volume option, and not part of 'all'. This translator is making glusterfs's 'ctx' specific options as a possibility.

The one with 'all' is as mentioned by @kshlm are global for the cluster (and may be for glusterd itself)

**Comment by atinmu** [comment]
For all global cluster specific options like op-version, brick-multiplex et all are only limited to glusterd. Not sure if having a different ReST API would make much sense here as of users it's still to be done through a volume set interface, but I agree that we shouldn't try to put the logic of this segregation in the main volume set function. Would it make much sense to have something like:

gluster volume set cluster <global option> <value> for the ones which are related to cluster level tuning?

**Comment by aravindavk** [comment]
We need to clearly differentiate between the global options per volume(via global xlator) and Cluster wide options. 

As discussed with @prashanthpai and @amarts, Global options per Volume can be set using the same interface available today for other Volume options. (May need some changes to parse global.so file). For example, `glustercli volume set <volname> global.<opt-name> <value>`. This needs to be added to all generated Volfiles.

Setting Cluster wide options: Who will use these options? will it be added to volfile? or If it is default option which can be configured before any Volume Create, so that all new Volume create will have these options set by default. @atinmu @amarts

**Comment by atinmu** [comment]
> Setting Cluster wide options: Who will use these options? will it be added to volfile? or If it is default option which can be configured before any Volume Create, so that all new Volume create will have these options set by default. @atinmu @amarts

These options will not be written to any volfiles as the attributes are applicable at cluster level and they are purely a cluster wide tuning, they don't apply to a volume level and limited to glusterd only. As suggested by Kaushal, we just need to have these values stored in etcd.

**Comment by samikshan** [comment]
Right now https://github.com/gluster/glusterfs/blob/e7ee4d8949b015a1421d4d97b763ccb4dcfc6eb1/cli/src/cli-cmd-parser.c#L410 contains the list of volume names that aren't allowed currently. Implementing something like `volume set cluster` to set cluster level options, would add "cluster" to that list of volume names. Are we okay with that kind of a change?

The other option which is to have separate APIs/commands for setting/getting these options could be implemented. However, I am unsure about what data points should be present in the cluster object other than these options. Peer list seems like one of the data points but it would be redundant given that the peer list is already maintained.

File: glusterd2/commands/global/commands.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+			Name:        "SetGlobalOptions",
+			Method:      "POST",
+			Pattern:     "/cluster/options",
+			Version:     1,
+			HandlerFunc: setGlobalOptionsHandler,
+		},
+		route.Route{
+			Name:        "GetGlobalOptions",
+			Method:      "GET",
+			Pattern:     "/cluster/options/{optname}",
```
Comment (aravindavk) [inline glusterd2/commands/global/commands.go]: Return all options at once, CLI can filter if required. (optname can be removed)

File: glusterd2/commands/global/getglobaloptions.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/gorilla/mux"
+)
+
+func getGlobalOptionsHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	optname := mux.Vars(r)["optname"]
+
+	var c cluster.Cluster
+	resp, err := store.Store.Get(context.TODO(), cluster.ClusterPrefix)
```
Comment (aravindavk) [inline glusterd2/commands/global/getglobaloptions.go]: Create this as function in separate file `store-utils.go`. So that all store related functions will be in one place.

File: glusterd2/commands/global/setglobaloptions.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	ctx := r.Context()
+
+	var req api.GlobalOptionReq
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusUnprocessableEntity, errors.ErrJSONParsingFailed.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	var c cluster.Cluster
+	resp, err := store.Store.Get(context.TODO(), cluster.ClusterPrefix)
```
Comment (aravindavk) [inline glusterd2/commands/global/setglobaloptions.go]: Same as above comment, keep store related things in a separate file.

File: glusterd2/commands/global/setglobaloptions.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+			c.Options[k] = v
+		} else {
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, fmt.Sprintf("Invalid global option: %s", k), api.ErrCodeDefault)
+			continue
+		}
+	}
+
+	data, _ := json.Marshal(c)
+	if _, err := store.Store.Put(context.TODO(), cluster.ClusterPrefix, string(data)); err != nil {
```
Comment (aravindavk) [inline glusterd2/commands/global/setglobaloptions.go]: same

File: glusterd2/commands/global/setglobaloptions.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		return
+	}
+
+	if err = json.Unmarshal(resp.Kvs[0].Value, &c); err != nil {
+		return
+	}
+
+	for k, v := range req.Options {
+		if _, found := cluster.GlobalOptMap[k]; found {
+			// TODO validate the value type for global option
```
Comment (aravindavk) [inline glusterd2/commands/global/setglobaloptions.go]: Is it possible to utilize the volume options validation functions here?

File: glusterd2/commands/volumes/common.go [hunk 6]
```diff
@@ -18,7 +18,6 @@ import (
 // validateOptions validates if the options and their values are valid and can
 // be set on a volume.
 func validateOptions(opts map[string]string) error {
-
```
Comment (aravindavk) [inline glusterd2/commands/volumes/common.go]: nit: change not related to this patch

File: glusterd2/main.go [hunk 7]
```diff
@@ -108,6 +109,10 @@ func main() {
 		log.WithError(err).Fatal("Failed to generate local auth token")
 	}
 
+	if err := cluster.LoadClusterAttributes(); err != nil {
```
Comment (aravindavk) [inline glusterd2/main.go]: Not really required. since all the default values and other details available with Glusterd2 itself.

File: glusterd2/xlator/global.go [hunk 8]
```diff
@@ -14,7 +14,6 @@ var (
 
 // Load load all available xlators and intializes the xlators and options maps
 func Load() (err error) {
-
```
Comment (aravindavk) [inline glusterd2/xlator/global.go]: nit: unrelated

File: glusterd2/xlator/global.go [hunk 9]
```diff
@@ -37,6 +36,7 @@ func Xlators() map[string]*Xlator {
 
 // loadOptions loads all available options into the options.Options map,
 // indexed as <xlator-id>.<option-key> for all available option keys
+// Also initialise the globalOptMap with defauls/value from etcd server
```
Comment (aravindavk) [inline glusterd2/xlator/global.go]: Is this comment belong to previous patchset?

File: pkg/restclient/volume.go [hunk 10]
```diff
@@ -69,6 +69,13 @@ func (c *Client) VolumeSet(volname string, req api.VolOptionReq) error {
 	return err
 }
 
+// GlobalOptionSet sets cluster level options
+func (c *Client) GlobalOptionSet(req api.GlobalOptionReq) error {
+	url := fmt.Sprintf("/v1/cluster/options")
+	err := c.post(url, req, http.StatusOK, nil)
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: skip err variable.

    return c.post(url, req, http.StatusOK, nil)

File: glusterd2/cluster/cluster.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+// GlobalOption reperesents cluster wide options
+type GlobalOption struct {
+	Key          string
+	DefaultValue string
+	Type         options.OptionType
+}
+
+// LoadClusterAttributes updates etcd with cluster attributes if not already present
+func LoadClusterAttributes() error {
```
Comment (aravindavk) [inline glusterd2/cluster/cluster.go]: This function is not required.  Store only modified/set global options.

File: glusterd2/commands/global/getglobaloptions.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+	"github.com/gluster/glusterd2/glusterd2/cluster"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/store"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/gluster/glusterd2/pkg/errors"
+	"github.com/gorilla/mux"
+)
+
+func getGlobalOptionsHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (aravindavk) [inline glusterd2/commands/global/getglobaloptions.go]: Read default options to temp opts, then get from store and update temp opts and return. Good to return the following details.

     Name
     Value
     Default Value
     ModifiedFlag

File: pkg/restclient/volume.go [hunk 13]
```diff
@@ -69,6 +69,13 @@ func (c *Client) VolumeSet(volname string, req api.VolOptionReq) error {
...
+	return err
+}
+
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: Also a func required to get Global options

File: glusterd2/commands/global/setglobaloptions.go (hunk 5 above)
Comment (samikshan) [reply glusterd2/commands/global/setglobaloptions.go]: The validation functions for volume options will work only on the separately defined option type in xlator/options/options.go. In fact Validate() is a method and works only on volume options for now. I do feel that they should be made reusable for cluster wide option validation.

File: glusterd2/commands/global/getglobaloptions.go (hunk 2 above)
Comment (samikshan) [reply glusterd2/commands/global/getglobaloptions.go]: Sure.

File: glusterd2/xlator/global.go (hunk 9 above)
Comment (samikshan) [reply glusterd2/xlator/global.go]: Yes it is! Will remove.

File: glusterd2/commands/global/getglobaloptions.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+	"github.com/gluster/glusterd2/glusterd2/cluster"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/pkg/api"
+)
+
+func getGlobalOptionsHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	c, _ := cluster.GetCluster()
```
Comment (aravindavk) [inline glusterd2/commands/global/getglobaloptions.go]: Don't ignore error. 

Ignore only specific error if required. Error may be due to etcd connectivity.

File: pkg/restclient/volume.go [hunk 15]
```diff
@@ -69,6 +69,12 @@ func (c *Client) VolumeSet(volname string, req api.VolOptionReq) error {
 	return err
 }
 
+// GlobalOptionSet sets cluster level options
+func (c *Client) GlobalOptionSet(req api.GlobalOptionReq) error {
+	url := fmt.Sprintf("/v1/cluster/options")
+	return c.post(url, req, http.StatusOK, nil)
+}
+
```
Comment (aravindavk) [inline pkg/restclient/volume.go]: Also need `GlobalOptions()`

Comment (samikshan) [reply pkg/restclient/volume.go]: If you mean GlobalOptionsGet() do you think it makes sense to add this part after the patch on volume get support is merged?

File: glusterd2/commands/global/getglobaloptions.go (hunk 14 above)
Comment (samikshan) [reply glusterd2/commands/global/getglobaloptions.go]: Yes. Missed the error handling here.

File: pkg/restclient/volume.go (hunk 15 above)
Comment (aravindavk) [reply pkg/restclient/volume.go]: Sure, makes sense.

**Comment by samikshan** [comment]
Global options reset can definitely be done. I would prefer to work on it in a later patch though.
